File: 101-200/101-110/108-sortedArrayToBST/sortedArrayToBST.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        
        self.nums = nums
        return self.sortedArrayToBSTHelper(0, len(nums))
        
    def sortedArrayToBSTHelper(self, start, end):
        if start == end:
            return None
        
        mid = (start + end) / 2
        node = TreeNode(self.nums[mid])
        node.left = self.sortedArrayToBSTHelper(start, mid)
        node.right = self.sortedArrayToBSTHelper(mid+1, end)
        return node

File: 101-200/101-110/108-sortedArrayToBST/sortedArrayToBST.cpp
#include <vector>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

using namespace std;

class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return sortedArrayToBSTHelper(nums, 0, nums.size());
    }

    TreeNode* sortedArrayToBSTHelper(vector<int> &nums, int begin, int end) {
        if (begin == end)
            return NULL;
            
        int middle = (begin + end) / 2;
        TreeNode *root = new TreeNode(nums[middle]);
        root->left = sortedArrayToBSTHelper(nums, begin, middle);
        root->right = sortedArrayToBSTHelper(nums, middle+1, end);
        return root;
    }
};